src/main/java/io/cdap/plugin/gcp/dataplex/source/DataplexBatchSource.java

```diff
@@ -104,106 +124,319 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+        String.format("Read from BigQuery Entity %s '%s' from Dataplex.",
+          type, table));
+    }
   }
 
+  private String createTask(String locationId, String lakeId, String outputLocation, String query)
+    throws IOException, DataplexException {
+    String taskArgs = "--output_location,%s, --output_format, %s";
+    Task task = new Task();
+    Task.TriggerSpec triggerSpec = new Task.TriggerSpec();
+    triggerSpec.setType("ON_DEMAND");
+
+    Task.ExecutionSpec executionSpec = new Task.ExecutionSpec();
+    Task.Args args = new Task.Args();
+    executionSpec.setServiceAccount(config.getServiceAccountEmail());
+    args.setTaskArgs(String.format(taskArgs, outputLocation, DataplexBatchSourceConfig.INPUT_FORMAT));
+    executionSpec.setArgs(args);
+
+    Task.Spark spark = new Task.Spark();
+    spark.setSqlScript(query);
+    task.setTriggerSpec(triggerSpec);
+    task.setExecutionSpec(executionSpec);
+    task.setSpark(spark);
+    task.setDescription("task-" + UUID.randomUUID());
+    dataplexInterface.createTask(config.getCredentials(),
+      config.tryGetProject(), locationId, lakeId, task);
+    return task.getDescription();
+  }
+
+  private String formatQuery(Entity entityBean, boolean isPreviewEnabled) {
+    String queryTemplate = "select * from %s.%s %s";
+    StringBuilder condition = new StringBuilder();
+    if (!Strings.isNullOrEmpty(config.getFilter())) {
+      condition.append("where ").append(config.getFilter());
+    }
+    if (isPreviewEnabled) {
+      condition.append(" LIMIT 1000;");
+    }
+    return String.format(queryTemplate, config.getZone(), entityBean.getId(), condition);
+  }
+
+  /**
+   * Gets bucket from supplied configuration.
+   * <p>
+   * If the supplied configuration doesn't specify a bucket, a bucket will get auto created and configuration modified
+   * to auto-delete this bucket on completion.
+   *
+   * @param configuration Hadoop configuration instance.
+   * @param project       GCP projectId
+   * @param bigQuery      bigquery client
+   * @param credentials   GCP credentials
+   * @param bucketName    bucket Name
+   * @return Bucket name.
+   */
+  private String getOrCreateBucket(Configuration configuration,
+                                   String project,
+                                   BigQuery bigQuery,
+                                   Credentials credentials,
+                                   String bucketName) {
+    String bucket = String.format(BQ_TEMP_BUCKET_NAME_TEMPLATE, bucketName);
+    // By default, this option is false, meaning the job can not delete the bucket. So enable it only when bucket name
+    // is not provided.
+    configuration.setBoolean("fs.gs.bucket.delete.enable", true);
+
+    // the dataset existence is validated before, so this cannot be null
+    Dataset bigQueryDataset = bigQuery.getDataset(DatasetId.of(datasetProject, dataset));
+    GCPUtils.createBucket(GCPUtils.getStorage(project, credentials),
```

Comment:
Looks like we're always creating a bucket, is that intentional?

Reply:
Yes, we will always create a bucket.

Reply:
then we should rename the method. getOrCreateBucket() is not accurate